walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,60 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
+  /**Return collection of timeRanges that work given the events all invited attendees are going to.*/
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> sortedEvents = new ArrayList<Event>(events);
+    Collections.sort(sortedEvents, Event.ORDER_BY_START);
+
+    List<TimeRange> availableTimes = new ArrayList<TimeRange>();
+    Set<String> requestAttendees = new HashSet<String>(request.getAttendees());
+    TimeRange prevEventTime = null;
+    for (Event currEvent: sortedEvents) {
+      Set<String> currAttendees = currEvent.getAttendees();
+      if (Collections.disjoint(requestAttendees, currAttendees)) {
+        continue;
+      } 
+      TimeRange currEventTime = currEvent.getWhen();
+      if (prevEventTime != null && prevEventTime.overlaps(currEventTime)) {
+        prevEventTime = combineOverlappingRanges(prevEventTime, currEventTime);
+      } else {
+        int start = (prevEventTime == null) ? TimeRange.START_OF_DAY : prevEventTime.end();
+        int end = currEventTime.start();
+        availableTimes = addTime(availableTimes, start, end, request.getDuration());
+        prevEventTime = currEventTime;
+      }
+    }
+
+    if (prevEventTime != null) {
+      availableTimes = addTime(availableTimes, prevEventTime.end(), TimeRange.END_OF_DAY, request.getDuration());
+    } else {
+      availableTimes = addTime(availableTimes, TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, request.getDuration());
+    }
+
+    return availableTimes;
+  }
+
+  /** Create one big range out of two overlapping ones. */
+  private TimeRange combineOverlappingRanges(TimeRange range1, TimeRange range2) {
+    int start = Math.min(range1.start(), range2.start());
+    int end = Math.max(range1.end(), range2.end());
+    return TimeRange.fromStartEnd(start, end, false);
+  }
+
+  /** Adds available time in a failsafe way. */
+  private List<TimeRange> addTime(List<TimeRange> lst, int start, int end, long minDuration) {
+    boolean inclusive = (end == TimeRange.END_OF_DAY) ? true : false;
+    if (start < end && end - start >= minDuration && start >= TimeRange.START_OF_DAY && end <= TimeRange.END_OF_DAY) {
```

Comment:
This should be factored out as sth like hasSufficientTime, and it's clearer to separate control and execution. Concretely, a Do
if (condition) { Do() }
is usually better than a MaybeDo
MaybeDoSomething { if (condition) {do something} }.